### PR TITLE
Update Vectorization.mdx

### PR DESCRIPTION
**Pull request by Alexis-Tao88**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [X ] I have tested my code.
- [X ] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X ] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: content/6_Advanced/Vectorization.mdx [hunk 1]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
 <!-- Anyways, keep in mind that these rarely make a difference in competitive programming. -->
 
 <!-- <IncompleteSection /> -->
+
+## Example - Ligatures
+
+[Problem](https://open.kattis.com/problems/ligatures?editresubmit=13798177)
+We have yet to find a problem on this topic. In the meantime, use this.
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:140]: wdym by this

File: content/6_Advanced/Vectorization.mdx [hunk 2]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
...
+
+The problem involves counting the occurrences of specific pairs of characters (ligatures) in a given text corpus based on multiple sets of suggested ligatures. The primary challenge is efficiently handling large input sizes and ensuring that the counting process respects the rule that ligatures can't overlap.
+
+### Solution
+
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:145]: please follow module layout here
https://docs.google.com/document/d/1Dzj5uYNYNXe_E-_9-lSeX73654cGJk88PZ0kZCCmks0/edit

File: content/6_Advanced/Vectorization.mdx [hunk 3]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
...
+- The core of the solution utilizes SIMD instructions to parallelize the counting process.
+The program processes the corpus in chunks, accumulating counts for each query simultaneously using AVX2 intrinsics.
+
+**Time Complexity:** $\mathcal{O}(N + Q * K)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#pragma GCC target("avx2")
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:166]: what's this

File: content/6_Advanced/Vectorization.mdx [hunk 4]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
...
+#include <vector>
+using namespace std;
+
+#define rep(i, from, to) for (int i = from; i < (to); ++i)
+#define trav(a, x) for (auto &a : x)
+#define all(x) x.begin(), x.end()
+#define sz(x) (int)(x).size()
+typedef long long ll;
+typedef pair<int, int> pii;
+typedef vector<int> vi;
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:179]: please don't use these

File: content/6_Advanced/Vectorization.mdx [hunk 5]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
...
+	u.m = m;
+	rep(i, 0, SIMD_BITS) { cout << ((u.ar[i / 32] >> (i % 32)) & 1); }
+	cout << endl;
+}
+
+mi active[MAXQ / SIMD_BITS + 1][26 * 26 + 1]{};  // ~26*26 * MAXQ / 8 bytes
+
+int main() {
+	cin.sync_with_stdio(false);
+	cin.exceptions(cin.failbit);
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:210]: what's this

File: content/6_Advanced/Vectorization.mdx [hunk 6]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
...
+	}  // (carry, acc) = (t + acc + carry)
+#define ADD(ind) ADD2(t##ind, acc##ind)
+		while (i < padN) {
+			mi t0, t1, t2, t3, t4;
+			SET0;
+			t0 = carry;
+			SET0;
+			ADD(0);
+			t1 = carry;
+			SET0;
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:259]: what the heck

File: content/6_Advanced/Vectorization.mdx [hunk 7]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
...
+			rep(l, 0, SIMD_BITS / 32) {
+				res[k * SIMD_BITS + l * 32 + j] = u.ar[l];
+			}
+		}
+	}
+
+	rep(i, 0, Q) cout << res[i] << '\n';
+	cout << flush;
+
+	exit(0);
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:392]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
is this for bounties

**Comment by ryanchou-dev** [comment]
> is this for bounties

isn't on the sheet

File: content/6_Advanced/Vectorization.mdx (hunk 3 above)
Comment (ryanchou-dev) [reply content/6_Advanced/Vectorization.mdx:166]: i think these optimizations are the point of the module (pragmas)

File: content/6_Advanced/Vectorization.mdx (hunk 6 above)
Comment (ryanchou-dev) [reply content/6_Advanced/Vectorization.mdx:259]: in reference to this: https://en.wikipedia.org/wiki/Streaming_SIMD_Extensions from the module?

**Comment by ryanchou-dev** [comment]
@bqi343

hey! if you're free, could you review the content for this module?

File: content/6_Advanced/Vectorization.mdx (hunk 4 above)
Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx:179]: it's fine

File: content/6_Advanced/Vectorization.mdx [hunk 8]
```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
 <!-- Anyways, keep in mind that these rarely make a difference in competitive programming. -->
 
 <!-- <IncompleteSection /> -->
+
+## Example - Ligatures
```
Comment (bqi343) [inline content/6_Advanced/Vectorization.mdx:137]: this should be a practice problem with an internal solution. see the practice problems with internal solutions at the end of other modules for examples

File: content/6_Advanced/Vectorization.mdx (hunk 5 above)
Comment (Alexis-Tao88) [reply content/6_Advanced/Vectorization.mdx:210]: This code snippet sets up the environment for using SIMD (Single Instruction, Multiple Data) operations with AVX2 instructions in C++. 
The function _set_bit_ sets a specific bit in a 256-bit vector (_mi_ type) at the given index (_ind_). The function _printm_  prints the binary representation of a 256-bit vector (_mi_ type).
In the main function, _cin.sync_with_stdio(false)_  disables the synchronization between C++ standard streams and C standard streams, which improves I/O performance. _cin.exceptions(cin.failbit)_ configures the input stream to throw exceptions if it encounters an input failure, enabling more robust error handling.

File: content/6_Advanced/Vectorization.mdx (hunk 3 above)
Comment (Alexis-Tao88) [reply content/6_Advanced/Vectorization.mdx:166]: This is just a breakdown of the solution and what it accomplishes.

File: content/6_Advanced/Vectorization.mdx (hunk 1 above)
Comment (Alexis-Tao88) [reply content/6_Advanced/Vectorization.mdx:140]: The module states that this concept has not appeared in a problem yet. The content author who brought up this issue linked an example problem that wasn't from USACO, so I wasn't sure how to format this section.

File: content/6_Advanced/Vectorization.mdx (hunk 6 above)
Comment (Alexis-Tao88) [reply content/6_Advanced/Vectorization.mdx:259]: Sorry, this was the solution that the content author linked to.I apologize for not commenting beforehand. 
Here is my understanding:
It prepares the data structures (pairs array, active array) and initializes N, Q, and K for processing ligature sets using SIMD instructions. It uses padding for alignment, vectorized operations, and direct memory manipulation for performance optimizations. This setup is important for the following SIMD-based computations to efficiently count ligature occurrences in the given corpus and ligature sets.

File: content/6_Advanced/Vectorization.mdx (hunk 1 above)
Comment (ryanchou-dev) [reply content/6_Advanced/Vectorization.mdx:140]: this could probably be omitted - the module already mentions this 👍

**Review by ryanchou-dev — APPROVED** [review]
if you could add some more comments to the code, that'd be great. otherwise, looks good (formatting-wise).

im not sure how many people on our team can review this module, but we should fine merging it once kevin approves

File: content/6_Advanced/Vectorization.mdx (hunk 8 above)
Comment (Alexis-Tao88) [reply content/6_Advanced/Vectorization.mdx:137]: sorry abt that, I have created a second pull request in a new mdx file in the solutions division

Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx:137]: can you add that to this pull request? should be simple if you're using Github Desktop

File: content/6_Advanced/Vectorization.mdx [hunk 9]
```diff
@@ -133,3 +133,275 @@ instead of faster (ex. see
...
+
+<FocusProblem problem="tutorial" />
+
+[Problem](https://open.kattis.com/problems/ligatures?editresubmit=13798177)
+
+### Explanation
+
+The problem involves counting the occurrences of specific pairs of characters (ligatures) in a given text corpus based on multiple sets of suggested ligatures. The primary challenge is efficiently handling large input sizes and ensuring that the counting process respects the rule that ligatures can't overlap.
+
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:146]:
```suggestion
```

File: content/6_Advanced/Vectorization.mdx [hunk 10]
```diff
@@ -133,3 +133,275 @@ instead of faster (ex. see
...
+### Solution
+
+This solution to the problem uses SIMD (Single Instruction, Multiple Data) parallelism to efficiently count the occurrences of suggested ligatures within a large corpus:
```
Comment (SansPapyrus683) [inline content/6_Advanced/Vectorization.mdx:149]: i mean is there any guide you can give on how people can implement something similar themselves?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Vectorization.mdx (hunk 8 above)
Comment (Alexis-Tao88) [reply content/6_Advanced/Vectorization.mdx:137]: oky the answer is now in the solutions division in this pull request

File: content/6_Advanced/Vectorization.mdx [hunk 11]
```diff
@@ -133,3 +133,269 @@ instead of faster (ex. see
...
+
+2. **Setting Up Bitmasks**:
+- A bitmask array active is set up where each bit represents whether a specific ligature is part of a query.
+The bitmask for each query is configured such that each suggested ligature toggles a specific bit.
+
+3. **Counting Occurrences Using SIMD**:
+- The core of the solution utilizes SIMD instructions to parallelize the counting process.
+The program processes the corpus in chunks, accumulating counts for each query simultaneously using AVX2 intrinsics.
+
+**Time Complexity:** $\mathcal{O}(N + Q * K)$
```
Comment (bqi343) [inline content/6_Advanced/Vectorization.mdx:157]: this complexity is wrong. it's actually Q(N+K) (with a really good constant factor for the QN part).

File: content/6_Advanced/Vectorization.mdx (hunk 8 above)
Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx:137]: ok, then can you remove the solution from the module? and make sure the solution is linked from the module (same as internal solution for any other problem)

File: content/6_Advanced/Vectorization.mdx (hunk 10 above)
Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx:147]: The solution code is rather cryptic, but if you understand roughly what it is doing then you should be able to write your own solution along the same lines. Though the current solution description is rather vague so it's hard to understand what the code is doing. Here's my understanding.

The idea is to first write the naive $NQ$ solution, then figure out how to speed it up by processing batches of 256 queries in parallel. Naive solution code:

```
int main() {
	setIO();
	auto to_int = [&](char a, char b) {
		return 26 * (a - 'a') + (b - 'a');
	};  // map pair to index in [0, 676)
	def(int, N, Q, K);
	def(str, corpus);
	V<V<bool>> active(676, V<bool>(Q));
	F0R(q, Q) {
		def(str, ligature);
		for (int i = 0; i < K; ++i)
			active[to_int(ligature[2 * i], ligature[2 * i + 1])][q] = true;
	}
	F0R(q, Q) {
		int counter = 0;
		bool last = 0;
		FOR(i, 1, N) {
			last = !last & active[to_int(corpus[i - 1], corpus[i])][q];
			counter += last;
		}
		ps(counter);
	}
}
```

Now consider what needs to change if we want to process 256 queries in parallel. First, we can change `last` from a single bit to a 256-bit integer type (`__m256i` in the solution code). We can do the same with `active`.

Parallelizing the `counter += last` step is trickier. It is represented by `accs` and `ts` in the code. Basically, counter is a 20-bit integer, and we will have two 256-bit integers for each bit in counter, such that `counter = accs + ts`. To output `counter` at the end we first need to add `accs` and ts` (the `// Final accumulation across all indices` part).

File: content/6_Advanced/Vectorization.problems.json [hunk 12]
```diff
@@ -1,3 +1,17 @@
...
+  "MODULE_ID": "vectorization",
+  "probs": [
+    {
+      "uniqueId": "Ligatures",
+      "name": "Ligatures",
+      "url": "https://open.kattis.com/problems/ligatures?editresubmit=13798177",
+      "source": "Advanced",
+      "difficulty": "Hard",
+      "isStarred": false,
+      "tags": [""],
```
Comment (bqi343) [inline content/6_Advanced/Vectorization.problems.json:11]:
```suggestion
      "tags": [],
```

File: content/6_Advanced/Vectorization.mdx (hunk 10 above)
Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx]: my solution code along these lines. omits optimizations such as the manual loop unrolling, but still passes just under the TL

```cpp
#pragma GCC target("avx2")
#include <immintrin.h>

using mi = __m256i;
const int SIMD_BITS = 256;

void set_bit(mi &m, size_t ind) {
	union {
		mi m;
		uint32_t ar[SIMD_BITS / 32];
	} u;
	u.m = m;
	u.ar[ind / 32] = (uint32_t)(u.ar[ind / 32] | (1 << (ind % 32)));
	m = u.m;
}

int main() {
	setIO();
	def(int, N, Q, K);
	def(str, corpus);
	vs ligs(Q);
	re(ligs);
	vi ans(Q);
	auto to_int = [&](char a, char b) { return 26 * (a - 'a') + (b - 'a'); };
	for (int ql = 0; ql < Q;) {
		int qr = min(Q, ql + SIMD_BITS);
		mi active[676]{};
		FOR(q, ql, qr)
		F0R(k, K)
		set_bit(active[to_int(ligs[q].at(2 * k), ligs[q].at(2 * k + 1))],
		        q - ql);
		mi bits_a[20]{}, bits_b[20]{};  // counter = bits_a + bits_b
		mi last{};
		FOR(i, 1, N) {
			int pair = to_int(corpus.at(i - 1), corpus.at(i));
			// update last
			last = ~last & active[pair];
			// update counter
			mi cur = last;
			for (int pos = 0;; ++pos) {  // this loop runs O(1) times on average
				if (!(i & (1 << pos))) {
					bits_b[pos] = cur;
					break;
				}
				if ((1 << (pos + 1)) > i) {
					bits_a[pos] = cur;
					break;
				}
				// reduce three bits to two (1, 1, 1 -> 1, 2)
				// bits_a, bits_b, cur -> res0, res1
				auto res0 = bits_a[pos] ^ bits_b[pos] ^ cur;
				auto res1 = (bits_a[pos] & bits_b[pos]) |
				            (cur & (bits_a[pos] | bits_b[pos]));
				bits_a[pos] = res0;
				bits_b[pos] = _mm256_setzero_si256();
				cur = res1;
			}
		}
		F0R(pos, 20) {  // extract counter from bits_a and bits_b
			union {
				mi m;
				uint32_t ar[SIMD_BITS / 32];
			} u;
			u.m = bits_a[pos];
			FOR(q, ql, qr) {
				int idx = q - ql;
				if (u.ar[idx / 32] & (1 << (idx % 32))) { ans[q] += 1 << pos; }
			}
			u.m = bits_b[pos];
			FOR(q, ql, qr) {
				int idx = q - ql;
				if (u.ar[idx / 32] & (1 << (idx % 32))) { ans[q] += 1 << pos; }
			}
		}
		ql = qr;
	}
	each(t, ans) ps(t);
	// you should actually read the stuff at the bottom
}
```

Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx]: also, some relevant resources about SIMD: https://codeforces.com/blog/entry/94152?#comment-1162758

File: content/6_Advanced/Vectorization.problems.json [hunk 13]
```diff
@@ -1,3 +1,17 @@
 {
-  "MODULE_ID": "vectorization"
+  "MODULE_ID": "vectorization",
+  "probs": [
+    {
+      "uniqueId": "Ligatures",
+      "name": "Ligatures",
+      "url": "https://open.kattis.com/problems/ligatures?editresubmit=13798177",
+      "source": "Advanced",
```
Comment (bqi343) [inline content/6_Advanced/Vectorization.problems.json:8]:
```suggestion
      "source": "Kattis",
```

File: content/6_Advanced/Vectorization.problems.json [hunk 14]
```diff
@@ -1,3 +1,17 @@
 {
-  "MODULE_ID": "vectorization"
+  "MODULE_ID": "vectorization",
+  "probs": [
+    {
+      "uniqueId": "Ligatures",
+      "name": "Ligatures",
+      "url": "https://open.kattis.com/problems/ligatures?editresubmit=13798177",
```
Comment (bqi343) [inline content/6_Advanced/Vectorization.problems.json:7]:
```suggestion
      "url": "https://open.kattis.com/problems/ligatures",
```

File: content/6_Advanced/Vectorization.problems.json [hunk 15]
```diff
@@ -1,3 +1,17 @@
 {
-  "MODULE_ID": "vectorization"
+  "MODULE_ID": "vectorization",
+  "probs": [
+    {
+      "uniqueId": "Ligatures",
```
Comment (bqi343) [inline content/6_Advanced/Vectorization.problems.json:5]:
```suggestion
      "uniqueId": "kattis-ligatures",
```

File: content/6_Advanced/Vectorization.mdx (hunk 10 above)
Comment (bqi343) [reply content/6_Advanced/Vectorization.mdx]: @Alexis-Tao88 good to link this in the internal sol

File: solutions/advanced/Ligatures.mdx [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+- The corpus is preprocessed to identify all pairs of consecutive characters, which are stored in a pairs vector.
+
+2. **Setting Up Bitmasks**:
+- A bitmask array active is set up where each bit represents whether a specific ligature is part of a query.
+The bitmask for each query is configured such that each suggested ligature toggles a specific bit.
+
+3. **Counting Occurrences Using SIMD**:
+- The core of the solution utilizes SIMD instructions to parallelize the counting process.
+The program processes the corpus in chunks, accumulating counts for each query simultaneously using AVX2 intrinsics.
+
```
Comment (bqi343) [inline solutions/advanced/Ligatures.mdx:27]: add `<IncompleteSection />` around here given that this solution explanation is by no means complete.

File: solutions/advanced/Ligatures.mdx [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+### Explanation
+
+The problem involves counting the occurrences of specific pairs of characters (ligatures) in a given text corpus based on multiple sets of suggested ligatures. The primary challenge is efficiently handling large input sizes and ensuring that the counting process respects the rule that ligatures can't overlap.
+
+### Solution
+
+This solution to the problem uses SIMD (Single Instruction, Multiple Data) parallelism to efficiently count the occurrences of suggested ligatures within a large corpus:
+
+**Overview**
+1. **Preprocessing the Corpus**:
```
Comment (bqi343) [inline solutions/advanced/Ligatures.mdx:17]: the numbering is messed up - you probably want to add spaces before the hyphens to fix this

Screenshot 2024-06-23 at 7 11 22 PM

File: solutions/advanced/Ligatures.mdx [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+	carry = picked = _mm256_andnot_si256(picked, *(mi *)(act + pairs[i++]))
+#define ADD2(t, acc)                                                           \
+	{                                                                          \
+		mi x = t ^ carry, y = (x & acc) | (t & carry);                         \
+		acc = x ^ acc, carry = y;                                              \
+	}  // (carry, acc) = (t + acc + carry)
+#define ADD(ind) ADD2(t##ind, acc##ind)
+
+		// Loop to process the pairs and accumulate the results
+		while (i < padN) {
```
Comment (bqi343) [inline solutions/advanced/Ligatures.mdx:128]: remove the loop unrolling - the solution still runs fast enough without it, and it adds 100 lines to the code length

it should suffice to remove acc0 ... acc4 and start `ind` at 0

Comment (bqi343) [reply solutions/advanced/Ligatures.mdx:128]: after this you should also be able to remove `padN`

**Review by bqi343 — CHANGES_REQUESTED** [review]
please link the corresponding github issue. after all changes are addressed, should be good to merge

File: solutions/advanced/Ligatures.mdx (hunk 17 above)
Comment (bqi343) [reply solutions/advanced/Ligatures.mdx:17]: you can check in https://usaco.guide/editor/

**Comment by ryanchou-dev** [comment]
Closes #2477

File: solutions/advanced/Ligatures.mdx [hunk 19]
```diff
@@ -0,0 +1,171 @@
+---
+id: kattis-ligatures
+source: Kattis
+title: Ligatures
+author: Alexis Tao
+---
+
+### Explanation
+
+The problem involves counting the occurrences of specific pairs of characters (ligatures) in a given text corpus based on multiple sets of suggested ligatures. The primary challenge is efficiently handling large input sizes and ensuring that the counting process respects the rule that ligatures can't overlap.
```
Comment (SansPapyrus683) [inline solutions/advanced/Ligatures.mdx:10]: is there a focus problem link

Comment (bqi343) [reply solutions/advanced/Ligatures.mdx:10]: not a focus problem, but it's linked here: https://usaco-guide-18mueqdak-cpinitiative.vercel.app/adv/vectorization#problems

**Review by bqi343 — APPROVED** [review]

File: solutions/advanced/Ligatures.mdx (hunk 19 above)
Comment (SansPapyrus683) [reply solutions/advanced/Ligatures.mdx:10]: i mean if it's the problem the module goes over...shouldn't it be put as one instead?

Comment (bqi343) [reply solutions/advanced/Ligatures.mdx:10]: it's not supposed to be an in-module problem

Comment (SansPapyrus683) [reply solutions/advanced/Ligatures.mdx:10]: wait how does that even work

Comment (bqi343) [reply solutions/advanced/Ligatures.mdx:10]: how does what work?

Comment (SansPapyrus683) [reply solutions/advanced/Ligatures.mdx:10]: when would a problem not be in-module? i don't think any other module has this kind of thing...

Comment (bqi343) [reply solutions/advanced/Ligatures.mdx:10]: I don't think there's any competitive programming problem like this one 🤷. It feels strange to add it in-module with no examples after it.

**Pull request merged**